src/sentry/digests/backends/redis.py

```diff
@@ -72,30 +72,57 @@ def make_record_key(timeline_key, record):
     return '{0}:{1}:{2}'.format(timeline_key, TIMELINE_RECORD_PATH_COMPONENT, record)
 
 
-# Ensures an timeline is scheduled to be digested.
-# KEYS: {WATING, READY}
-# ARGV: {TIMELINE, TIMESTAMP}
+# Ensures an timeline is scheduled to be digested, adjusting the schedule time
+# if necessary.
+# KEYS: {WAITING, READY}
+# ARGV: {
+#   TIMELINE,   -- timeline key
+#   TIMESTAMP,  --
+#   INCREMENT,  -- amount of time (in seconds) that an event addition delays scheduling
+#   MAXIMUM     -- maximum amount of time (in seconds) between a timeline being
+#               -- digested, and the same timeline being scheduled for the next
+#               -- digestion
+# }
 ENSURE_TIMELINE_SCHEDULED_SCRIPT = """\
--- Check to see if the timeline exists in the "waiting" set (heuristics tell us
--- that this should be more likely than it's presence in the "ready" set.)
-local waiting = redis.call('ZSCORE', KEYS[1], ARGV[1])
-
-if waiting ~= false then
-    -- If the item already exists, update the score if the provided timestamp
-    -- is less than the current score.
-    if tonumber(waiting) > tonumber(ARGV[2]) then
-        redis.call('ZADD', KEYS[1], ARGV[2], ARGV[1])
-    end
-    return
+-- If the timeline is already in the "ready" set, this is a noop.
+if tonumber(redis.call('ZSCORE', KEYS[2], ARGV[1])) ~= nil then
+    return false
 end
 
--- Otherwise, check to see if the timeline already exists in the "ready" set.
--- If it doesn't, it needs to be added to the "waiting" set to be scheduled.
-if redis.call('ZSCORE', KEYS[2], ARGV[1]) == false then
-    redis.call('ZADD', KEYS[1], ARGV[2], ARGV[1])
-    return
+-- Otherwise, check to see if the timeline is in the "waiting" set.
+local score = tonumber(redis.call('ZSCORE', KEYS[1], ARGV[1]))
+if score ~= nil then
+    -- If the timeline is already in the "waiting" set, increase the delay by
+    -- min(current schedule + increment value, maximum delay after last processing time).
+    local last = tonumber(redis.call('GET', ARGV[1] .. ':{TIMELINE_LAST_PROCESSED_TIMESTAMP_PATH_COMPONENT}'))
+    local update = nil;
+    if last == nil then
+        -- If the last processed timestamp is missing for some reason (possibly
+        -- evicted), be conservative and allow the timeline to be scheduled
+        -- with either the current schedule time or provided timestamp,
+        -- whichever is smaller.
+        update = math.min(last, ARGV[2])
```

Comment:
This block also handles backwards compatibility for the existing backoff implementation.